<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#SQL-Introduction" data-toc-modified-id="SQL-Introduction-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>SQL Introduction</a></span></li><li><span><a href="#Connect-the-sql" data-toc-modified-id="Connect-the-sql-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Connect the sql</a></span></li><li><span><a href="#Know-your-database" data-toc-modified-id="Know-your-database-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Know your database</a></span></li><li><span><a href="#Read-dat-files-of-database" data-toc-modified-id="Read-dat-files-of-database-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Read dat files of database</a></span><ul class="toc-item"><li><span><a href="#Get-pandas-dataframes" data-toc-modified-id="Get-pandas-dataframes-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Get pandas dataframes</a></span></li></ul></li><li><span><a href="#Section-2" data-toc-modified-id="Section-2-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Section 2</a></span><ul class="toc-item"><li><span><a href="#qn1" data-toc-modified-id="qn1-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>qn1</a></span></li><li><span><a href="#qn2" data-toc-modified-id="qn2-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>qn2</a></span></li><li><span><a href="#qn3" data-toc-modified-id="qn3-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>qn3</a></span></li><li><span><a href="#qn" data-toc-modified-id="qn-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>qn</a></span></li><li><span><a href="#qn" data-toc-modified-id="qn-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>qn</a></span></li></ul></li><li><span><a href="#Section-3:-Aggregation" data-toc-modified-id="Section-3:-Aggregation-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Section 3: Aggregation</a></span><ul class="toc-item"><li><span><a href="#Challanges" data-toc-modified-id="Challanges-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Challanges</a></span><ul class="toc-item"><li><span><a href="#qn1" data-toc-modified-id="qn1-6.1.1"><span class="toc-item-num">6.1.1&nbsp;&nbsp;</span>qn1</a></span></li><li><span><a href="#qn2" data-toc-modified-id="qn2-6.1.2"><span class="toc-item-num">6.1.2&nbsp;&nbsp;</span>qn2</a></span></li><li><span><a href="#qn3" data-toc-modified-id="qn3-6.1.3"><span class="toc-item-num">6.1.3&nbsp;&nbsp;</span>qn3</a></span></li></ul></li><li><span><a href="#Having-clause" data-toc-modified-id="Having-clause-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Having clause</a></span><ul class="toc-item"><li><span><a href="#Challanges" data-toc-modified-id="Challanges-6.2.1"><span class="toc-item-num">6.2.1&nbsp;&nbsp;</span>Challanges</a></span><ul class="toc-item"><li><span><a href="#qn1" data-toc-modified-id="qn1-6.2.1.1"><span class="toc-item-num">6.2.1.1&nbsp;&nbsp;</span>qn1</a></span></li><li><span><a href="#qn2" data-toc-modified-id="qn2-6.2.1.2"><span class="toc-item-num">6.2.1.2&nbsp;&nbsp;</span>qn2</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Section4:-Assessment-1" data-toc-modified-id="Section4:-Assessment-1-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Section4: Assessment 1</a></span><ul class="toc-item"><li><span><a href="#qn1" data-toc-modified-id="qn1-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>qn1</a></span></li><li><span><a href="#qn2" data-toc-modified-id="qn2-7.2"><span class="toc-item-num">7.2&nbsp;&nbsp;</span>qn2</a></span></li><li><span><a href="#qn3" data-toc-modified-id="qn3-7.3"><span class="toc-item-num">7.3&nbsp;&nbsp;</span>qn3</a></span></li></ul></li><li><span><a href="#Section5:-SQL-Joins" data-toc-modified-id="Section5:-SQL-Joins-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Section5: SQL Joins</a></span><ul class="toc-item"><li><span><a href="#AS" data-toc-modified-id="AS-8.1"><span class="toc-item-num">8.1&nbsp;&nbsp;</span>AS</a></span></li><li><span><a href="#JOIN" data-toc-modified-id="JOIN-8.2"><span class="toc-item-num">8.2&nbsp;&nbsp;</span>JOIN</a></span><ul class="toc-item"><li><span><a href="#INNER-JOIN" data-toc-modified-id="INNER-JOIN-8.2.1"><span class="toc-item-num">8.2.1&nbsp;&nbsp;</span>INNER JOIN</a></span><ul class="toc-item"><li><span><a href="#example-2" data-toc-modified-id="example-2-8.2.1.1"><span class="toc-item-num">8.2.1.1&nbsp;&nbsp;</span>example 2</a></span></li><li><span><a href="#example-3" data-toc-modified-id="example-3-8.2.1.2"><span class="toc-item-num">8.2.1.2&nbsp;&nbsp;</span>example 3</a></span></li><li><span><a href="#example-4" data-toc-modified-id="example-4-8.2.1.3"><span class="toc-item-num">8.2.1.3&nbsp;&nbsp;</span>example 4</a></span></li></ul></li></ul></li></ul></li></ul></div>

# SQL Introduction

![](../images/sql_clauses.png)
![](../images/sql_mnemonic.png)

# Connect the sql

In [1]:
import numpy as np
import pandas as pd
import os
import yaml

with open( os.path.expanduser('~') + "/.postgres_conf.yml", 'r') as stream:
    try:
        yaml_dict = yaml.safe_load(stream)
    except yaml.YAMLError as exc:
        print(exc)

pw = yaml_dict['password']
port = yaml_dict['port']

In [2]:
%load_ext sql

In [3]:
%sql postgres://postgres:$pw@localhost:$port/dvdrental

'Connected: postgres@dvdrental'

# Know your database

In [4]:
%%sql
SELECT TABLE_NAME
FROM INFORMATION_SCHEMA.TABLES
WHERE TABLE_TYPE = 'BASE TABLE'
AND TABLE_CATALOG='dvdrental'
and TABLE_NAME  not like 'pg_%'
and TABLE_NAME  not like 'sql_%'
-- order by table_name

 * postgres://postgres:***@localhost:5432/dvdrental
15 rows affected.


[('actor',),
 ('store',),
 ('address',),
 ('category',),
 ('city',),
 ('country',),
 ('customer',),
 ('film_actor',),
 ('film_category',),
 ('inventory',),
 ('language',),
 ('rental',),
 ('staff',),
 ('payment',),
 ('film',)]

In [5]:
%%sql
select column_name, data_type, character_maximum_length
    from INFORMATION_SCHEMA.COLUMNS 
    where table_name = 'customer';

 * postgres://postgres:***@localhost:5432/dvdrental
10 rows affected.


[('customer_id', 'integer', None),
 ('store_id', 'smallint', None),
 ('first_name', 'character varying', 45),
 ('last_name', 'character varying', 45),
 ('email', 'character varying', 50),
 ('address_id', 'smallint', None),
 ('activebool', 'boolean', None),
 ('create_date', 'date', None),
 ('last_update', 'timestamp without time zone', None),
 ('active', 'integer', None)]

In [6]:
%%sql
SELECT TABLE_NAME
FROM INFORMATION_SCHEMA.TABLES
WHERE TABLE_TYPE = 'BASE TABLE'
AND TABLE_CATALOG='dvdrental'
and TABLE_NAME  not like 'pg_%'
and TABLE_NAME  not like 'sql_%'
-- order by table_name

 * postgres://postgres:***@localhost:5432/dvdrental
15 rows affected.


[('actor',),
 ('store',),
 ('address',),
 ('category',),
 ('city',),
 ('country',),
 ('customer',),
 ('film_actor',),
 ('film_category',),
 ('inventory',),
 ('language',),
 ('rental',),
 ('staff',),
 ('payment',),
 ('film',)]

In [7]:
%%sql
-- select * from staff limit 2; -- This gives error in jupyter notebook
select * from actor limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(1, 'Penelope', 'Guiness', datetime.datetime(2013, 5, 26, 14, 47, 57, 620000)),
 (2, 'Nick', 'Wahlberg', datetime.datetime(2013, 5, 26, 14, 47, 57, 620000))]

# Read dat files of database

In [8]:
!ls ../data/dvdrental/

2163.dat    2167.dat    2169.dat    2173.dat    2177.dat    2181.dat    2185.dat    2189.dat    toc.dat
2165.dat    2168.dat    2171.dat    2175.dat    2179.dat    2183.dat    2187.dat    restore.sql


In [9]:
!head ../data/dvdrental/2163.dat

1	Penelope	Guiness	2013-05-26 14:47:57.62
2	Nick	Wahlberg	2013-05-26 14:47:57.62
3	Ed	Chase	2013-05-26 14:47:57.62
4	Jennifer	Davis	2013-05-26 14:47:57.62
5	Johnny	Lollobrigida	2013-05-26 14:47:57.62
6	Bette	Nicholson	2013-05-26 14:47:57.62
7	Grace	Mostel	2013-05-26 14:47:57.62
8	Matthew	Johansson	2013-05-26 14:47:57.62
9	Joe	Swank	2013-05-26 14:47:57.62
10	Christian	Gable	2013-05-26 14:47:57.62


In [10]:
%%bash

for f in `ls ../data/dvdrental/[^toc]*.dat`                                                                
   do
      echo
      echo  "File: $f"; 
      echo "==================================================================="
      head -2  "$f";
done;


File: ../data/dvdrental/2163.dat
1	Penelope	Guiness	2013-05-26 14:47:57.62
2	Nick	Wahlberg	2013-05-26 14:47:57.62

File: ../data/dvdrental/2165.dat
1	Action	2006-02-15 09:46:27
2	Animation	2006-02-15 09:46:27

File: ../data/dvdrental/2167.dat
133	Chamber Italian	A Fateful Reflection of a Moose And a Husband who must Overcome a Monkey in Nigeria	2006	1	7	4.99	117	14.99	NC-17	2013-05-26 14:50:58.951	{Trailers}	'chamber':1 'fate':4 'husband':11 'italian':2 'monkey':16 'moos':8 'must':13 'nigeria':18 'overcom':14 'reflect':5
384	Grosse Wonderful	A Epic Drama of a Cat And a Explorer who must Redeem a Moose in Australia	2006	1	5	4.99	49	19.99	R	2013-05-26 14:50:58.951	{"Behind the Scenes"}	'australia':18 'cat':8 'drama':5 'epic':4 'explor':11 'gross':1 'moos':16 'must':13 'redeem':14 'wonder':2

File: ../data/dvdrental/2168.dat
1	1	2006-02-15 10:05:03
1	23	2006-02-15 10:05:03

File: ../data/dvdrental/2169.dat
1	6	2006-02-15 10:07:09
2	11	2006-02-15 10:07:09

File: ../data/dvdrental/2171.dat

## Get pandas dataframes
- 2187.dat staff
- 2165.dat category
- 2179.dat film_category
- 2175.dat country
- 2163.dat actor
- 2181.dat language
- 2179.dat inventory
- 2183.dat payment
- 2185.dat rental
- 2173.dat city
- 2179.dat store
- 2167.dat  film
- 2171.dat address
- 2168.dat film_actor
- 2177.dat customer

In [11]:
tables = ['staff', 'category', 'film_category', 'country', 'actor',
          'language', 'inventory', 'payment', 'rental', 'city',
          'store', 'film', 'address', 'film_actor', 'customer']

staff = tables[0]
category = tables[1]
film_category = tables[2]
country = tables[3]
actor = tables[4]
language = tables[5]
inventory = tables[6]
payment = tables[7]
rental = tables[8]
city = tables[9]
store = tables[10]
film = tables[11]
address = tables[12]
film_actor = tables[13]
customer = tables[14]

In [12]:
# staff = %sql select * from $staff;
# staff = customer.DataFrame()


In [13]:
category = %sql select * from $category;
category = category.DataFrame()

 * postgres://postgres:***@localhost:5432/dvdrental
16 rows affected.


In [14]:
# staff = %sql select * from $staff;  # this fails

film_category = %sql select * from $film_category;
country = %sql select * from $country;
actor = %sql select * from $actor;
language = %sql select * from $language;
inventory = %sql select * from $inventory;
payment = %sql select * from $payment;
rental = %sql select * from $rental;
city = %sql select * from $city;
store = %sql select * from $store;
film = %sql select * from $film;
address = %sql select * from $address;
film_actor = %sql select * from $film_actor;
customer = %sql select * from $customer;

 * postgres://postgres:***@localhost:5432/dvdrental
1000 rows affected.
 * postgres://postgres:***@localhost:5432/dvdrental
109 rows affected.
 * postgres://postgres:***@localhost:5432/dvdrental
200 rows affected.
 * postgres://postgres:***@localhost:5432/dvdrental
6 rows affected.
 * postgres://postgres:***@localhost:5432/dvdrental
4581 rows affected.
 * postgres://postgres:***@localhost:5432/dvdrental
14596 rows affected.
 * postgres://postgres:***@localhost:5432/dvdrental
16044 rows affected.
 * postgres://postgres:***@localhost:5432/dvdrental
600 rows affected.
 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.
 * postgres://postgres:***@localhost:5432/dvdrental
1000 rows affected.
 * postgres://postgres:***@localhost:5432/dvdrental
603 rows affected.
 * postgres://postgres:***@localhost:5432/dvdrental
5462 rows affected.
 * postgres://postgres:***@localhost:5432/dvdrental
599 rows affected.


In [15]:
film_category = film_category.DataFrame()
country = country.DataFrame()
actor = actor.DataFrame()
language = language.DataFrame()
inventory = inventory.DataFrame()
payment = payment.DataFrame()
rental = rental.DataFrame()
city = city.DataFrame()
store = store.DataFrame()
film = film.DataFrame()
address = address.DataFrame()
film_actor = film_actor.DataFrame()
customer = customer.DataFrame()

In [16]:
# I got memory error for city table of file 2173.dat
# I will manually read the datafile using pandas

In [17]:
# I can not open staff table in jupyter notebook, but I can open this in
# Query tool of pgAdmin.

"""
select column_name from INFORMATION_SCHEMA.COLUMNS 
where table_name = 'customer';


Gives:

"staff_id"
"first_name"
"last_name"
"address_id"
"email"
"store_id"
"active"
"username"
"password"
"last_update"
"picture"


"""

'\nselect column_name from INFORMATION_SCHEMA.COLUMNS \nwhere table_name = \'customer\';\n\n\nGives:\n\n"staff_id"\n"first_name"\n"last_name"\n"address_id"\n"email"\n"store_id"\n"active"\n"username"\n"password"\n"last_update"\n"picture"\n\n\n'

In [18]:
cols = ['staff_id', 'first_name', 'last_name', 'address_id', 'email',
        'store_id', 'active', 'username', 'password', 'last_update',
        'picture']

In [19]:
!head ../data/dvdrental/2187.dat

1	Mike	Hillyer	3	Mike.Hillyer@sakilastaff.com	1	t	Mike	8cb2237d0679ca88db6464eac60da96345513964	2006-05-16 16:13:11.79328	\\x89504e470d0a5a0a
2	Jon	Stephens	4	Jon.Stephens@sakilastaff.com	2	t	Jon	8cb2237d0679ca88db6464eac60da96345513964	2006-05-16 16:13:11.79328	\N




In [20]:
staff = pd.read_csv('../data/dvdrental/2187.dat', sep=r'\t', header=None, engine='python')
staff.columns = cols
staff = staff.head(2)
staff['active'] = True
print(staff.shape)
staff.head()

# looking at pgAdmin this table has only two rows and last column is nulls.
# active column t mean true.
# 

(2, 11)


staff_id first_name last_name  address_id                         email  \
0         1       Mike   Hillyer           3  Mike.Hillyer@sakilastaff.com   
1         2        Jon  Stephens           4  Jon.Stephens@sakilastaff.com   

   store_id  active username                                  password  \
0         1    True     Mike  8cb2237d0679ca88db6464eac60da96345513964   
1         2    True      Jon  8cb2237d0679ca88db6464eac60da96345513964   

                 last_update              picture  
0  2006-05-16 16:13:11.79328  \\x89504e470d0a5a0a  
1  2006-05-16 16:13:11.79328                   \N

In [21]:
df_tables = [staff, category, film_category, country, actor,
            language, inventory, payment, rental, city,
            store, film, address, film_actor, customer]

In [22]:
all_columns = [df.columns.tolist() for df in df_tables]
all_columns[0]

['staff_id',
 'first_name',
 'last_name',
 'address_id',
 'email',
 'store_id',
 'active',
 'username',
 'password',
 'last_update',
 'picture']

In [23]:
df_tables_cols = pd.DataFrame(all_columns).T.fillna('')

df_tables_cols.columns = tables
df_tables_cols

staff     category film_category      country        actor  \
0      staff_id  category_id       film_id   country_id     actor_id   
1    first_name         name   category_id      country   first_name   
2     last_name  last_update   last_update  last_update    last_name   
3    address_id                                          last_update   
4         email                                                        
5      store_id                                                        
6        active                                                        
7      username                                                        
8      password                                                        
9   last_update                                                        
10      picture                                                        
11                                                                     
12                                                                     

       language     inventory       payment        rental         city  \
0   language_id  inventory_id    payment_id     rental_id      city_id   
1          name       film_id   customer_id   rental_date         city   
2   last_update      store_id      staff_id  inventory_id   country_id   
3                 last_update     rental_id   customer_id  last_update   
4                                    amount   return_date                
5                              payment_date      staff_id                
6                                             last_update                
7                                                                        
8                                                                        
9                                                                        
10                                                                       
11                                                                       
12                                                                       

               store              film      address   film_actor     customer  
0           store_id           film_id   address_id     actor_id  customer_id  
1   manager_staff_id             title      address      film_id     store_id  
2         address_id       description     address2  last_update   first_name  
3        last_update      release_year     district                 last_name  
4                          language_id      city_id                     email  
5                      rental_duration  postal_code                address_id  
6                          rental_rate        phone                activebool  
7                               length  last_update               create_date  
8                     replacement_cost                            last_update  
9                               rating                                 active  
10                         last_update                                         
11                    special_features                                         
12                            fulltext

# Section 2

## qn1
How many payment transactions were greater than $5 ?

In [24]:
%%sql
select * from customer limit 5;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[(524, 1, 'Jared', 'Ely', 'jared.ely@sakilacustomer.org', 530, True, datetime.date(2006, 2, 14), datetime.datetime(2013, 5, 26, 14, 49, 45, 738000), 1),
 (1, 1, 'Mary', 'Smith', 'mary.smith@sakilacustomer.org', 5, True, datetime.date(2006, 2, 14), datetime.datetime(2013, 5, 26, 14, 49, 45, 738000), 1),
 (2, 1, 'Patricia', 'Johnson', 'patricia.johnson@sakilacustomer.org', 6, True, datetime.date(2006, 2, 14), datetime.datetime(2013, 5, 26, 14, 49, 45, 738000), 1),
 (3, 1, 'Linda', 'Williams', 'linda.williams@sakilacustomer.org', 7, True, datetime.date(2006, 2, 14), datetime.datetime(2013, 5, 26, 14, 49, 45, 738000), 1),
 (4, 2, 'Barbara', 'Jones', 'barbara.jones@sakilacustomer.org', 8, True, datetime.date(2006, 2, 14), datetime.datetime(2013, 5, 26, 14, 49, 45, 738000), 1)]

In [25]:
%%sql
select * from payment limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(17503, 341, 2, 1520, Decimal('7.99'), datetime.datetime(2007, 2, 15, 22, 25, 46, 996577)),
 (17504, 341, 1, 1778, Decimal('1.99'), datetime.datetime(2007, 2, 16, 17, 23, 14, 996577))]

In [26]:
%%sql
select count(*) from payment
where amount > 5;

 * postgres://postgres:***@localhost:5432/dvdrental
1 rows affected.


[(3618,)]

In [27]:
payment.query('amount > 5').size # ignores nulls

21708

In [28]:
payment.query('amount > 5').shape[0]

3618

## qn2
How many actors have a first name that starts with the letter P?


In [29]:
%%sql
select * from actor limit 5;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[(1, 'Penelope', 'Guiness', datetime.datetime(2013, 5, 26, 14, 47, 57, 620000)),
 (2, 'Nick', 'Wahlberg', datetime.datetime(2013, 5, 26, 14, 47, 57, 620000)),
 (3, 'Ed', 'Chase', datetime.datetime(2013, 5, 26, 14, 47, 57, 620000)),
 (4, 'Jennifer', 'Davis', datetime.datetime(2013, 5, 26, 14, 47, 57, 620000)),
 (5, 'Johnny', 'Lollobrigida', datetime.datetime(2013, 5, 26, 14, 47, 57, 620000))]

In [30]:
%%sql
select count(*) from actor
where first_name like 'P%';

 * postgres://postgres:***@localhost:5432/dvdrental
1 rows affected.


[(5,)]

In [31]:
actor[actor.first_name.str.startswith('P')].shape

(5, 4)

In [32]:
actor[actor.first_name.str.startswith('P')]

actor_id first_name last_name             last_update
0           1   Penelope   Guiness 2013-05-26 14:47:57.620
45         46     Parker  Goldberg 2013-05-26 14:47:57.620
53         54   Penelope   Pinkett 2013-05-26 14:47:57.620
104       104   Penelope    Cronyn 2013-05-26 14:47:57.620
120       120   Penelope    Monroe 2013-05-26 14:47:57.620

## qn3
How many unique districts are our customers from?


In [33]:
df_tables_cols

staff     category film_category      country        actor  \
0      staff_id  category_id       film_id   country_id     actor_id   
1    first_name         name   category_id      country   first_name   
2     last_name  last_update   last_update  last_update    last_name   
3    address_id                                          last_update   
4         email                                                        
5      store_id                                                        
6        active                                                        
7      username                                                        
8      password                                                        
9   last_update                                                        
10      picture                                                        
11                                                                     
12                                                                     

       language     inventory       payment        rental         city  \
0   language_id  inventory_id    payment_id     rental_id      city_id   
1          name       film_id   customer_id   rental_date         city   
2   last_update      store_id      staff_id  inventory_id   country_id   
3                 last_update     rental_id   customer_id  last_update   
4                                    amount   return_date                
5                              payment_date      staff_id                
6                                             last_update                
7                                                                        
8                                                                        
9                                                                        
10                                                                       
11                                                                       
12                                                                       

               store              film      address   film_actor     customer  
0           store_id           film_id   address_id     actor_id  customer_id  
1   manager_staff_id             title      address      film_id     store_id  
2         address_id       description     address2  last_update   first_name  
3        last_update      release_year     district                 last_name  
4                          language_id      city_id                     email  
5                      rental_duration  postal_code                address_id  
6                          rental_rate        phone                activebool  
7                               length  last_update               create_date  
8                     replacement_cost                            last_update  
9                               rating                                 active  
10                         last_update                                         
11                    special_features                                         
12                            fulltext

In [34]:
%%sql
select * from address limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(1, '47 MySakila Drive', None, 'Alberta', 300, '', '', datetime.datetime(2006, 2, 15, 9, 45, 30)),
 (2, '28 MySQL Boulevard', None, 'QLD', 576, '', '', datetime.datetime(2006, 2, 15, 9, 45, 30))]

In [35]:
%%sql
select count(distinct(district)) from address;

 * postgres://postgres:***@localhost:5432/dvdrental
1 rows affected.


[(378,)]

In [36]:
address.district.nunique()

378

## qn
How many films have a rating of R and a replacement cost between `$5` and `$15` ?

In [37]:
%%sql
select * from film limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(133, 'Chamber Italian', 'A Fateful Reflection of a Moose And a Husband who must Overcome a Monkey in Nigeria', 2006, 1, 7, Decimal('4.99'), 117, Decimal('14.99'), 'NC-17', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'chamber':1 'fate':4 'husband':11 'italian':2 'monkey':16 'moos':8 'must':13 'nigeria':18 'overcom':14 'reflect':5"),
 (384, 'Grosse Wonderful', 'A Epic Drama of a Cat And a Explorer who must Redeem a Moose in Australia', 2006, 1, 5, Decimal('4.99'), 49, Decimal('19.99'), 'R', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Behind the Scenes'], "'australia':18 'cat':8 'drama':5 'epic':4 'explor':11 'gross':1 'moos':16 'must':13 'redeem':14 'wonder':2")]

In [38]:
%%sql
select count(*) from film
where rating = 'R' and replacement_cost between 5 and 15;

 * postgres://postgres:***@localhost:5432/dvdrental
1 rows affected.


[(52,)]

In [39]:
film.query("rating == 'R' and 5 <= replacement_cost <= 15 ").shape

(52, 13)

## qn
How many films have the word Truman somewhere in the title?



In [40]:
%%sql
select * from film limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(133, 'Chamber Italian', 'A Fateful Reflection of a Moose And a Husband who must Overcome a Monkey in Nigeria', 2006, 1, 7, Decimal('4.99'), 117, Decimal('14.99'), 'NC-17', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'chamber':1 'fate':4 'husband':11 'italian':2 'monkey':16 'moos':8 'must':13 'nigeria':18 'overcom':14 'reflect':5"),
 (384, 'Grosse Wonderful', 'A Epic Drama of a Cat And a Explorer who must Redeem a Moose in Australia', 2006, 1, 5, Decimal('4.99'), 49, Decimal('19.99'), 'R', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Behind the Scenes'], "'australia':18 'cat':8 'drama':5 'epic':4 'explor':11 'gross':1 'moos':16 'must':13 'redeem':14 'wonder':2")]

In [41]:
%%sql
select count(*) from film
where title like '%Truman%';

 * postgres://postgres:***@localhost:5432/dvdrental
1 rows affected.


[(5,)]

In [42]:
%%sql
select film_id, title from film
where title like '%Truman%';

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[(6, 'Agent Truman'),
 (537, 'Lover Truman'),
 (579, 'Minds Truman'),
 (613, 'Mystic Truman'),
 (915, 'Truman Crazy')]

In [43]:
film[film.title.str.contains('Truman')].iloc[:,:2]

film_id          title
9          6   Agent Truman
538      537   Lover Truman
579      579   Minds Truman
613      613  Mystic Truman
914      915   Truman Crazy

# Section 3: Aggregation

![](../images/sql_aggregation.png)
![](../images/where_having.png)
![](../images/sql-count-group-by.png)

In [44]:
%%sql
select round(avg(amount),2) from payment;

 * postgres://postgres:***@localhost:5432/dvdrental
1 rows affected.


[(Decimal('4.20'),)]

In [45]:
payment.amount.mean().round(2)

4.2

In [46]:
%%sql
select amount from payment
order by amount limit 3;

 * postgres://postgres:***@localhost:5432/dvdrental
3 rows affected.


[(Decimal('0.00'),), (Decimal('0.00'),), (Decimal('0.00'),)]

In [47]:
%%sql
select count(amount) from payment
where amount = 0.00;

 * postgres://postgres:***@localhost:5432/dvdrental
1 rows affected.


[(24,)]

In [48]:
payment.query('amount == 0.00').shape

(24, 6)

In [49]:
%%sql
select customer_id, sum(amount)
from payment
group by customer_id
order by sum(amount) desc
limit 5;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[(148, Decimal('211.55')),
 (526, Decimal('208.58')),
 (178, Decimal('194.61')),
 (137, Decimal('191.62')),
 (144, Decimal('189.60'))]

In [50]:
payment[['customer_id','amount']]\
.groupby('customer_id')['amount']\
.sum()\
.sort_values(ascending=False)\
.reset_index()\
.head()

customer_id  amount
0          148  211.55
1          526  208.58
2          178  194.61
3          137  191.62
4          144  189.60

In [51]:
 %%sql
select * from payment limit 5;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[(17503, 341, 2, 1520, Decimal('7.99'), datetime.datetime(2007, 2, 15, 22, 25, 46, 996577)),
 (17504, 341, 1, 1778, Decimal('1.99'), datetime.datetime(2007, 2, 16, 17, 23, 14, 996577)),
 (17505, 341, 1, 1849, Decimal('7.99'), datetime.datetime(2007, 2, 16, 22, 41, 45, 996577)),
 (17506, 341, 2, 2829, Decimal('2.99'), datetime.datetime(2007, 2, 19, 19, 39, 56, 996577)),
 (17507, 341, 2, 3130, Decimal('7.99'), datetime.datetime(2007, 2, 20, 17, 31, 48, 996577))]

In [52]:
%%sql
select staff_id, count(payment_id)
from payment
group by staff_id;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(1, 7292), (2, 7304)]

In [53]:
payment[['staff_id','payment_id']]\
.groupby('staff_id')\
.payment_id.count()

staff_id
1    7292
2    7304
Name: payment_id, dtype: int64

In [54]:
%%sql
select * from film limit 5;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[(133, 'Chamber Italian', 'A Fateful Reflection of a Moose And a Husband who must Overcome a Monkey in Nigeria', 2006, 1, 7, Decimal('4.99'), 117, Decimal('14.99'), 'NC-17', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'chamber':1 'fate':4 'husband':11 'italian':2 'monkey':16 'moos':8 'must':13 'nigeria':18 'overcom':14 'reflect':5"),
 (384, 'Grosse Wonderful', 'A Epic Drama of a Cat And a Explorer who must Redeem a Moose in Australia', 2006, 1, 5, Decimal('4.99'), 49, Decimal('19.99'), 'R', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Behind the Scenes'], "'australia':18 'cat':8 'drama':5 'epic':4 'explor':11 'gross':1 'moos':16 'must':13 'redeem':14 'wonder':2"),
 (8, 'Airport Pollock', 'A Epic Tale of a Moose And a Girl who must Confront a Monkey in Ancient India', 2006, 1, 6, Decimal('4.99'), 54, Decimal('15.99'), 'R', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'airport':1 'ancient':18 'confront':14 'epic':4 'girl':11 'india':19 'monkey':16 'moos':8 'must':13 'pollock':2 'tale':5"),
 (98, 'Bright Encounters', 'A Fateful Yarn of a Lumberjack And a Feminist who must Conquer a Student in A Jet Boat', 2006, 1, 4, Decimal('4.99'), 73, Decimal('12.99'), 'PG-13', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'boat':20 'bright':1 'conquer':14 'encount':2 'fate':4 'feminist':11 'jet':19 'lumberjack':8 'must':13 'student':16 'yarn':5"),
 (1, 'Academy Dinosaur', 'A Epic Drama of a Feminist And a Mad Scientist who must Battle a Teacher in The Canadian Rockies', 2006, 1, 6, Decimal('0.99'), 86, Decimal('20.99'), 'PG', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Deleted Scenes', 'Behind the Scenes'], "'academi':1 'battl':15 'canadian':20 'dinosaur':2 'drama':5 'epic':4 'feminist':8 'mad':11 'must':14 'rocki':21 'scientist':12 'teacher':17")]

In [55]:
%%sql
select rating, count(rating) from film
group by rating;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[('R', 195), ('NC-17', 210), ('G', 178), ('PG', 194), ('PG-13', 223)]

In [56]:
film.rating.value_counts()

PG-13    223
NC-17    210
R        195
PG       194
G        178
Name: rating, dtype: int64

In [57]:
film.rating.value_counts().reset_index()\
.rename(columns={'index': 'rating', 'rating':'count'})

rating  count
0  PG-13    223
1  NC-17    210
2      R    195
3     PG    194
4      G    178

In [58]:
%%sql
select rental_duration, count(rental_duration)
from film
group by rental_duration;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[(4, 203), (6, 212), (7, 191), (3, 203), (5, 191)]

In [59]:
film.rental_duration.value_counts()

6    212
4    203
3    203
7    191
5    191
Name: rental_duration, dtype: int64

In [60]:
%%sql
select rating, avg(rental_rate)
from film
group by rating;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[('R', Decimal('2.9387179487179487')),
 ('NC-17', Decimal('2.9709523809523810')),
 ('G', Decimal('2.8888764044943820')),
 ('PG', Decimal('3.0518556701030928')),
 ('PG-13', Decimal('3.0348430493273543'))]

In [61]:
film.head(2)

film_id             title  \
0      133   Chamber Italian   
1      384  Grosse Wonderful   

                                         description  release_year  \
0  A Fateful Reflection of a Moose And a Husband ...          2006   
1  A Epic Drama of a Cat And a Explorer who must ...          2006   

   language_id  rental_duration rental_rate  length replacement_cost rating  \
0            1                7        4.99     117            14.99  NC-17   
1            1                5        4.99      49            19.99      R   

              last_update     special_features  \
0 2013-05-26 14:50:58.951           [Trailers]   
1 2013-05-26 14:50:58.951  [Behind the Scenes]   

                                            fulltext  
0  'chamber':1 'fate':4 'husband':11 'italian':2 ...  
1  'australia':18 'cat':8 'drama':5 'epic':4 'exp...

In [62]:
film[['rating','rental_rate']].head(2)

rating rental_rate
0  NC-17        4.99
1      R        4.99

In [63]:
film[['rating','rental_rate']].dtypes

rating         object
rental_rate    object
dtype: object

In [64]:
film['rental_rate'] = pd.to_numeric(film.rental_rate, errors='coerce')

In [65]:
film[['rating','rental_rate']].dtypes

rating          object
rental_rate    float64
dtype: object

In [66]:
%%sql
select rating, avg(rental_rate)
from film
group by rating
order by rating;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[('G', Decimal('2.8888764044943820')),
 ('PG', Decimal('3.0518556701030928')),
 ('PG-13', Decimal('3.0348430493273543')),
 ('R', Decimal('2.9387179487179487')),
 ('NC-17', Decimal('2.9709523809523810'))]

In [67]:
film[['rating','rental_rate']].groupby('rating')['rental_rate'].mean().sort_index()

rating
G        2.888876
NC-17    2.970952
PG       3.051856
PG-13    3.034843
R        2.938718
Name: rental_rate, dtype: float64

## Challanges

### qn1
- We have two staff members with Staff IDs 1 and 2.
We want to give a bonus to the staff member that handled the most payments.
- How many payments did each staff member handle?
And how much was the total amount processed by each staff member

In [68]:
%%sql
select * from payment limit 5;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[(17503, 341, 2, 1520, Decimal('7.99'), datetime.datetime(2007, 2, 15, 22, 25, 46, 996577)),
 (17504, 341, 1, 1778, Decimal('1.99'), datetime.datetime(2007, 2, 16, 17, 23, 14, 996577)),
 (17505, 341, 1, 1849, Decimal('7.99'), datetime.datetime(2007, 2, 16, 22, 41, 45, 996577)),
 (17506, 341, 2, 2829, Decimal('2.99'), datetime.datetime(2007, 2, 19, 19, 39, 56, 996577)),
 (17507, 341, 2, 3130, Decimal('7.99'), datetime.datetime(2007, 2, 20, 17, 31, 48, 996577))]

In [69]:
%%sql
select staff_id, count(*), sum(amount) from payment
group by staff_id;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(1, 7292, Decimal('30252.12')), (2, 7304, Decimal('31059.92'))]

In [70]:
payment[['staff_id','amount']]\
.groupby('staff_id')\
.agg(['count','sum'])

amount          
          count       sum
staff_id                 
1          7292  30252.12
2          7304  31059.92

### qn2
- Corporate headquarters is auditing our store!
They want to know the average replacement
cost of movies by rating.
- For example, R rated movies have an average
replacement cost of $ 20.23

In [71]:
%%sql
select * from film limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(133, 'Chamber Italian', 'A Fateful Reflection of a Moose And a Husband who must Overcome a Monkey in Nigeria', 2006, 1, 7, Decimal('4.99'), 117, Decimal('14.99'), 'NC-17', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'chamber':1 'fate':4 'husband':11 'italian':2 'monkey':16 'moos':8 'must':13 'nigeria':18 'overcom':14 'reflect':5"),
 (384, 'Grosse Wonderful', 'A Epic Drama of a Cat And a Explorer who must Redeem a Moose in Australia', 2006, 1, 5, Decimal('4.99'), 49, Decimal('19.99'), 'R', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Behind the Scenes'], "'australia':18 'cat':8 'drama':5 'epic':4 'explor':11 'gross':1 'moos':16 'must':13 'redeem':14 'wonder':2")]

In [72]:
%%sql
select rating, avg(replacement_cost)
from film
group by rating;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[('R', Decimal('20.2310256410256410')),
 ('NC-17', Decimal('20.1376190476190476')),
 ('G', Decimal('20.1248314606741573')),
 ('PG', Decimal('18.9590721649484536')),
 ('PG-13', Decimal('20.4025560538116592'))]

In [73]:
film.head(2)

film_id             title  \
0      133   Chamber Italian   
1      384  Grosse Wonderful   

                                         description  release_year  \
0  A Fateful Reflection of a Moose And a Husband ...          2006   
1  A Epic Drama of a Cat And a Explorer who must ...          2006   

   language_id  rental_duration  rental_rate  length replacement_cost rating  \
0            1                7         4.99     117            14.99  NC-17   
1            1                5         4.99      49            19.99      R   

              last_update     special_features  \
0 2013-05-26 14:50:58.951           [Trailers]   
1 2013-05-26 14:50:58.951  [Behind the Scenes]   

                                            fulltext  
0  'chamber':1 'fate':4 'husband':11 'italian':2 ...  
1  'australia':18 'cat':8 'drama':5 'epic':4 'exp...

In [74]:
film.columns

Index(['film_id', 'title', 'description', 'release_year', 'language_id',
       'rental_duration', 'rental_rate', 'length', 'replacement_cost',
       'rating', 'last_update', 'special_features', 'fulltext'],
      dtype='object')

In [75]:
film[['rating','replacement_cost']].head(2)

rating replacement_cost
0  NC-17            14.99
1      R            19.99

In [76]:
film[['rating','replacement_cost']].dtypes

rating              object
replacement_cost    object
dtype: object

In [77]:
film['replacement_cost'] = pd.to_numeric(film['replacement_cost'], errors='coerce')

In [78]:
film[['rating','replacement_cost']].dtypes

rating               object
replacement_cost    float64
dtype: object

In [79]:
film[['rating','replacement_cost']].groupby('rating')['replacement_cost'].mean().round(2)

rating
G        20.12
NC-17    20.14
PG       18.96
PG-13    20.40
R        20.23
Name: replacement_cost, dtype: float64

In [80]:
%%sql
select rating, round(avg(replacement_cost),2)
from film
group by rating;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[('R', Decimal('20.23')),
 ('NC-17', Decimal('20.14')),
 ('G', Decimal('20.12')),
 ('PG', Decimal('18.96')),
 ('PG-13', Decimal('20.40'))]

### qn3
- We want to send coupons to the 5 customers
who have spent the most amount of money.
- Get me the customer ids of the top 5
spenders.

In [81]:
%%sql
select * from payment limit 5;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[(17503, 341, 2, 1520, Decimal('7.99'), datetime.datetime(2007, 2, 15, 22, 25, 46, 996577)),
 (17504, 341, 1, 1778, Decimal('1.99'), datetime.datetime(2007, 2, 16, 17, 23, 14, 996577)),
 (17505, 341, 1, 1849, Decimal('7.99'), datetime.datetime(2007, 2, 16, 22, 41, 45, 996577)),
 (17506, 341, 2, 2829, Decimal('2.99'), datetime.datetime(2007, 2, 19, 19, 39, 56, 996577)),
 (17507, 341, 2, 3130, Decimal('7.99'), datetime.datetime(2007, 2, 20, 17, 31, 48, 996577))]

In [82]:
%%sql
select  sum(amount) from payment
limit 5;

 * postgres://postgres:***@localhost:5432/dvdrental
1 rows affected.


[(Decimal('61312.04'),)]

In [83]:
%%sql
select column_name, data_type, character_maximum_length
    from INFORMATION_SCHEMA.COLUMNS 
    where table_name = 'payment';

 * postgres://postgres:***@localhost:5432/dvdrental
6 rows affected.


[('payment_id', 'integer', None),
 ('customer_id', 'smallint', None),
 ('staff_id', 'smallint', None),
 ('rental_id', 'integer', None),
 ('amount', 'numeric', None),
 ('payment_date', 'timestamp without time zone', None)]

In [84]:
%%sql
select customer_id
from payment limit 5;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[(341,), (341,), (341,), (341,), (341,)]

In [85]:
%%sql
-- select customer_id, sum(amount) from payment limit 5; -- fails

 * postgres://postgres:***@localhost:5432/dvdrental
(psycopg2.ProgrammingError) can't execute an empty query [SQL: '-- select customer_id, sum(amount) from payment limit 5; -- fails'] (Background on this error at: http://sqlalche.me/e/f405)


In [86]:
%%sql
select customer_id, sum(amount)
from payment
group by customer_id
order by sum(amount) desc
limit 5;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[(148, Decimal('211.55')),
 (526, Decimal('208.58')),
 (178, Decimal('194.61')),
 (137, Decimal('191.62')),
 (144, Decimal('189.60'))]

In [87]:
payment.dtypes

payment_id               int64
customer_id              int64
staff_id                 int64
rental_id                int64
amount                  object
payment_date    datetime64[ns]
dtype: object

In [88]:
payment['amount'] = pd.to_numeric(payment.amount, errors='coerce')

In [89]:
payment[['customer_id','amount']]\
.groupby('customer_id')['amount']\
.sum()\
.sort_values(ascending=False)\
.head()

customer_id
148    211.55
526    208.58
178    194.61
137    191.62
144    189.60
Name: amount, dtype: float64

In [90]:
payment[['customer_id','amount']]\
.groupby('customer_id')['amount']\
.sum()\
.nlargest(5)

customer_id
148    211.55
526    208.58
178    194.61
137    191.62
144    189.60
Name: amount, dtype: float64

## Having clause
```sql
SELECT column_name(s)
FROM table_name
WHERE condition
GROUP BY column_name(s)
HAVING condition
ORDER BY column_name(s);
```

![](../images/having.png)
![](../images/sql_mnemonic.png)

In [91]:
%%sql
select store_id, count(customer_id)
from customer
group by store_id;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(1, 326), (2, 273)]

In [92]:
%%sql
select store_id, count(customer_id)
from customer
group by store_id
having count(customer_id) > 300;

 * postgres://postgres:***@localhost:5432/dvdrental
1 rows affected.


[(1, 326)]

In [93]:
%%sql
select rating, avg(rental_rate)
from film
where rating in ('R','G','PG')
group by rating
having avg(rental_rate) < 3
order by avg(rental_rate) desc;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[('R', Decimal('2.9387179487179487')), ('G', Decimal('2.8888764044943820'))]

### Challanges

#### qn1
> We want to know what customers are eligible
for our platinum credit card. The requirements
are that the customer has at least a total of 40
transaction payments.

> What customers (by customer id) are eligible
for the credit card?

In [94]:
%%sql
select * from payment limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(17503, 341, 2, 1520, Decimal('7.99'), datetime.datetime(2007, 2, 15, 22, 25, 46, 996577)),
 (17504, 341, 1, 1778, Decimal('1.99'), datetime.datetime(2007, 2, 16, 17, 23, 14, 996577))]

In [95]:
%%sql
select customer_id, count(amount)
from payment
group by customer_id
having count(amount) >= 40;

 * postgres://postgres:***@localhost:5432/dvdrental
3 rows affected.


[(144, 40), (526, 42), (148, 45)]

In [96]:
payment[['customer_id','amount']]\
.groupby('customer_id')['amount']\
.count()\
.reset_index()\
.query('amount >= 40')

customer_id  amount
143          144      40
147          148      45
525          526      42

In [97]:
payment[['customer_id','amount']]\
.groupby('customer_id')['amount']\
.count()\
.loc[lambda x: x >= 40]

customer_id
144    40
148    45
526    42
Name: amount, dtype: int64

#### qn2
> When grouped by rating, what movie ratings
have an average rental duration of more than
5 days?

In [98]:
%%sql
select * from film limit 1;

 * postgres://postgres:***@localhost:5432/dvdrental
1 rows affected.


[(133, 'Chamber Italian', 'A Fateful Reflection of a Moose And a Husband who must Overcome a Monkey in Nigeria', 2006, 1, 7, Decimal('4.99'), 117, Decimal('14.99'), 'NC-17', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'chamber':1 'fate':4 'husband':11 'italian':2 'monkey':16 'moos':8 'must':13 'nigeria':18 'overcom':14 'reflect':5")]

In [99]:
%%sql
select rating, avg(rental_duration)
from film
group by rating
having avg(rental_duration) > 5;

 * postgres://postgres:***@localhost:5432/dvdrental
3 rows affected.


[('NC-17', Decimal('5.1428571428571429')),
 ('PG', Decimal('5.0824742268041237')),
 ('PG-13', Decimal('5.0538116591928251'))]

In [100]:
film[['rating','rental_duration']]\
.groupby('rating')['rental_duration']\
.mean().loc[lambda x: x > 5]

rating
NC-17    5.142857
PG       5.082474
PG-13    5.053812
Name: rental_duration, dtype: float64

# Section4: Assessment 1
ASSESSMENT TEST 1

COMPLETE THE FOLLOWING TASKS!

1. Return the customer IDs of customers who have spent at least $110 with the staff member who has an ID of 2.

The answer should be customers 187 and 148.

2. How many films begin with the letter J?

The answer should be 20.

3. What customer has the highest customer ID number whose name starts with an 'E' and has an address ID lower than 500?

The answer is Eddie Tomlin

## qn1
Return the customer IDs of customers who have spent at least $110 with the staff member who has an ID of 2.
The answer should be customers 187 and 148.

In [101]:
df_tables_cols

staff     category film_category      country        actor  \
0      staff_id  category_id       film_id   country_id     actor_id   
1    first_name         name   category_id      country   first_name   
2     last_name  last_update   last_update  last_update    last_name   
3    address_id                                          last_update   
4         email                                                        
5      store_id                                                        
6        active                                                        
7      username                                                        
8      password                                                        
9   last_update                                                        
10      picture                                                        
11                                                                     
12                                                                     

       language     inventory       payment        rental         city  \
0   language_id  inventory_id    payment_id     rental_id      city_id   
1          name       film_id   customer_id   rental_date         city   
2   last_update      store_id      staff_id  inventory_id   country_id   
3                 last_update     rental_id   customer_id  last_update   
4                                    amount   return_date                
5                              payment_date      staff_id                
6                                             last_update                
7                                                                        
8                                                                        
9                                                                        
10                                                                       
11                                                                       
12                                                                       

               store              film      address   film_actor     customer  
0           store_id           film_id   address_id     actor_id  customer_id  
1   manager_staff_id             title      address      film_id     store_id  
2         address_id       description     address2  last_update   first_name  
3        last_update      release_year     district                 last_name  
4                          language_id      city_id                     email  
5                      rental_duration  postal_code                address_id  
6                          rental_rate        phone                activebool  
7                               length  last_update               create_date  
8                     replacement_cost                            last_update  
9                               rating                                 active  
10                         last_update                                         
11                    special_features                                         
12                            fulltext

In [102]:
%%sql
select * from payment limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(17503, 341, 2, 1520, Decimal('7.99'), datetime.datetime(2007, 2, 15, 22, 25, 46, 996577)),
 (17504, 341, 1, 1778, Decimal('1.99'), datetime.datetime(2007, 2, 16, 17, 23, 14, 996577))]

In [103]:
%%sql
select customer_id, sum(amount)
from payment
where staff_id = 2
group by customer_id
having sum(amount) >= 110;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(187, Decimal('110.81')), (148, Decimal('110.78'))]

In [104]:
payment.query('staff_id == 2')[['customer_id','amount']]\
.groupby('customer_id').sum().reset_index().query('amount >= 110')

customer_id  amount
147          148  110.78
186          187  110.81

## qn2
How many films begin with the letter J?
The answer should be 20.

In [105]:
%%sql
select * from film limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(133, 'Chamber Italian', 'A Fateful Reflection of a Moose And a Husband who must Overcome a Monkey in Nigeria', 2006, 1, 7, Decimal('4.99'), 117, Decimal('14.99'), 'NC-17', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'chamber':1 'fate':4 'husband':11 'italian':2 'monkey':16 'moos':8 'must':13 'nigeria':18 'overcom':14 'reflect':5"),
 (384, 'Grosse Wonderful', 'A Epic Drama of a Cat And a Explorer who must Redeem a Moose in Australia', 2006, 1, 5, Decimal('4.99'), 49, Decimal('19.99'), 'R', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Behind the Scenes'], "'australia':18 'cat':8 'drama':5 'epic':4 'explor':11 'gross':1 'moos':16 'must':13 'redeem':14 'wonder':2")]

In [106]:
%%sql
select count(title) from film
where title like 'J%';

 * postgres://postgres:***@localhost:5432/dvdrental
1 rows affected.


[(20,)]

In [107]:
film.title.loc[film.title.str.startswith('J')].shape

(20,)

In [108]:
film.title.loc[film.title.str.startswith('J')]

472          Jacket Frisco
473             Jade Bunch
474           Japanese Run
475             Jason Trap
476    Jawbreaker Brooklyn
477             Jaws Harry
478           Jedi Beneath
479        Jeepers Wedding
480         Jekyll Frogmen
481        Jeopardy Encino
482          Jericho Mulan
483          Jerk Paycheck
484           Jersey Sassy
485          Jet Neighbors
486       Jingle Sagebrush
487         Joon Northwest
488         Juggler Hardly
489          Jumanji Blade
490          Jumping Wrath
491          Jungle Closer
Name: title, dtype: object

In [109]:
film.loc[film.title.str.startswith('J')][['title']]

title
472        Jacket Frisco
473           Jade Bunch
474         Japanese Run
475           Jason Trap
476  Jawbreaker Brooklyn
477           Jaws Harry
478         Jedi Beneath
479      Jeepers Wedding
480       Jekyll Frogmen
481      Jeopardy Encino
482        Jericho Mulan
483        Jerk Paycheck
484         Jersey Sassy
485        Jet Neighbors
486     Jingle Sagebrush
487       Joon Northwest
488       Juggler Hardly
489        Jumanji Blade
490        Jumping Wrath
491        Jungle Closer

In [110]:
film.title.filter(regex=r'472') # pandas series filter if for index filtering.

472    Jacket Frisco
Name: title, dtype: object

In [111]:
film.title.reset_index().set_index('title').filter(regex='^J.*',axis=0)

index
title                     
Jacket Frisco          472
Jade Bunch             473
Japanese Run           474
Jason Trap             475
Jawbreaker Brooklyn    476
Jaws Harry             477
Jedi Beneath           478
Jeepers Wedding        479
Jekyll Frogmen         480
Jeopardy Encino        481
Jericho Mulan          482
Jerk Paycheck          483
Jersey Sassy           484
Jet Neighbors          485
Jingle Sagebrush       486
Joon Northwest         487
Juggler Hardly         488
Jumanji Blade          489
Jumping Wrath          490
Jungle Closer          491

## qn3
> What customer has the highest customer ID number whose name starts with an 'E' and has an address ID lower than 500?

- The answer is Eddie Tomlin

In [112]:
df_tables_cols.style.apply(lambda x: ["background: salmon" if  
  (v == 'customer_id' or v == 'address_id') else "" for v in x], axis = 1)

In [113]:
%%sql
select * from customer limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(524, 1, 'Jared', 'Ely', 'jared.ely@sakilacustomer.org', 530, True, datetime.date(2006, 2, 14), datetime.datetime(2013, 5, 26, 14, 49, 45, 738000), 1),
 (1, 1, 'Mary', 'Smith', 'mary.smith@sakilacustomer.org', 5, True, datetime.date(2006, 2, 14), datetime.datetime(2013, 5, 26, 14, 49, 45, 738000), 1)]

In [114]:
%%sql
select customer_id, first_name, last_name, address_id
from customer
where first_name like 'E%' and address_id < 500
order by customer_id desc
limit 1;

 * postgres://postgres:***@localhost:5432/dvdrental
1 rows affected.


[(434, 'Eddie', 'Tomlin', 439)]

In [115]:
customer[customer.first_name.str.startswith('E')]\
.query('address_id < 500')\
.nlargest(1, 'customer_id')\
.loc[:, ['customer_id','first_name','last_name','address_id']]

customer_id first_name last_name  address_id
434          434      Eddie    Tomlin         439

In [116]:
customer[customer.first_name.str.startswith('E')]\
.loc[lambda x: x['address_id'] < 500]\
.nlargest(1, 'customer_id')\
.loc[:, ['customer_id','first_name','last_name','address_id']]

customer_id first_name last_name  address_id
434          434      Eddie    Tomlin         439

# Section5: SQL Joins

- [SQL JOINS Explained with Venn Diagrams](https://blog.codinghorror.com/a-visual-explanation-of-sql-joins/)
- [SQL JOIN Examples](http://www.sql-join.com/)
- [Wikipedia Page on SQL JOINS](https://www.wikiwand.com/en/Join_(SQL))

![](../images/sql_joins.png)

## AS

In [117]:
%%sql
select * from payment limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(17503, 341, 2, 1520, Decimal('7.99'), datetime.datetime(2007, 2, 15, 22, 25, 46, 996577)),
 (17504, 341, 1, 1778, Decimal('1.99'), datetime.datetime(2007, 2, 16, 17, 23, 14, 996577))]

In [118]:
%%sql
select payment_id as pid from payment limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(17503,), (17504,)]

## JOIN

[Pandas Comparison with SQL](https://pandas.pydata.org/pandas-docs/stable/getting_started/comparison/comparison_with_sql.html)

### INNER JOIN

In [119]:
%%sql
select * from customer limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(524, 1, 'Jared', 'Ely', 'jared.ely@sakilacustomer.org', 530, True, datetime.date(2006, 2, 14), datetime.datetime(2013, 5, 26, 14, 49, 45, 738000), 1),
 (1, 1, 'Mary', 'Smith', 'mary.smith@sakilacustomer.org', 5, True, datetime.date(2006, 2, 14), datetime.datetime(2013, 5, 26, 14, 49, 45, 738000), 1)]

In [120]:
%%sql
select * from payment limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(17503, 341, 2, 1520, Decimal('7.99'), datetime.datetime(2007, 2, 15, 22, 25, 46, 996577)),
 (17504, 341, 1, 1778, Decimal('1.99'), datetime.datetime(2007, 2, 16, 17, 23, 14, 996577))]

In [121]:
%%sql
-- only customer_id is in both tables, no need to do c.first_name, p.amount
select c.customer_id, first_name, last_name, email, amount, payment_date
from customer c
inner join payment p
on p.customer_id = c.customer_id
limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(524, 'Jared', 'Ely', 'jared.ely@sakilacustomer.org', Decimal('1.99'), datetime.datetime(2007, 2, 15, 8, 27, 50, 996577)),
 (524, 'Jared', 'Ely', 'jared.ely@sakilacustomer.org', Decimal('4.99'), datetime.datetime(2007, 2, 16, 7, 53, 4, 996577))]

In [122]:
%%sql
select c.customer_id, c.first_name, c.last_name, c.email, p.amount, p.payment_date
from customer c
inner join payment p
on p.customer_id = c.customer_id
limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(524, 'Jared', 'Ely', 'jared.ely@sakilacustomer.org', Decimal('1.99'), datetime.datetime(2007, 2, 15, 8, 27, 50, 996577)),
 (524, 'Jared', 'Ely', 'jared.ely@sakilacustomer.org', Decimal('4.99'), datetime.datetime(2007, 2, 16, 7, 53, 4, 996577))]

In [123]:
customer.merge(payment, on='customer_id').head(2)

customer_id  store_id first_name last_name                         email  \
0          524         1      Jared       Ely  jared.ely@sakilacustomer.org   
1          524         1      Jared       Ely  jared.ely@sakilacustomer.org   

   address_id  activebool create_date             last_update  active  \
0         530        True  2006-02-14 2013-05-26 14:49:45.738       1   
1         530        True  2006-02-14 2013-05-26 14:49:45.738       1   

   payment_id  staff_id  rental_id  amount               payment_date  
0       18202         1       1306    1.99 2007-02-15 08:27:50.996577  
1       18203         2       1651    4.99 2007-02-16 07:53:04.996577

In [124]:
pd.merge(customer, payment, on='customer_id').head(2)

customer_id  store_id first_name last_name                         email  \
0          524         1      Jared       Ely  jared.ely@sakilacustomer.org   
1          524         1      Jared       Ely  jared.ely@sakilacustomer.org   

   address_id  activebool create_date             last_update  active  \
0         530        True  2006-02-14 2013-05-26 14:49:45.738       1   
1         530        True  2006-02-14 2013-05-26 14:49:45.738       1   

   payment_id  staff_id  rental_id  amount               payment_date  
0       18202         1       1306    1.99 2007-02-15 08:27:50.996577  
1       18203         2       1651    4.99 2007-02-16 07:53:04.996577

In [125]:
# pandas also offer inner join between df1Col and df2Index
indexed_payment = payment.set_index('customer_id')
pd.merge(customer, indexed_payment, left_on='customer_id', right_index=True).head(2)

customer_id  store_id first_name last_name                         email  \
0          524         1      Jared       Ely  jared.ely@sakilacustomer.org   
0          524         1      Jared       Ely  jared.ely@sakilacustomer.org   

   address_id  activebool create_date             last_update  active  \
0         530        True  2006-02-14 2013-05-26 14:49:45.738       1   
0         530        True  2006-02-14 2013-05-26 14:49:45.738       1   

   payment_id  staff_id  rental_id  amount               payment_date  
0       18202         1       1306    1.99 2007-02-15 08:27:50.996577  
0       18203         2       1651    4.99 2007-02-16 07:53:04.996577

#### example 2

In [126]:
%%sql
select c.customer_id, first_name, last_name, email, amount, payment_date
from customer c
inner join payment p
on p.customer_id = c.customer_id
order by c.customer_id
limit 5;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[(1, 'Mary', 'Smith', 'mary.smith@sakilacustomer.org', Decimal('5.99'), datetime.datetime(2007, 2, 14, 23, 22, 38, 996577)),
 (1, 'Mary', 'Smith', 'mary.smith@sakilacustomer.org', Decimal('0.99'), datetime.datetime(2007, 2, 15, 16, 31, 19, 996577)),
 (1, 'Mary', 'Smith', 'mary.smith@sakilacustomer.org', Decimal('9.99'), datetime.datetime(2007, 2, 15, 19, 37, 12, 996577)),
 (1, 'Mary', 'Smith', 'mary.smith@sakilacustomer.org', Decimal('4.99'), datetime.datetime(2007, 2, 16, 13, 47, 23, 996577)),
 (1, 'Mary', 'Smith', 'mary.smith@sakilacustomer.org', Decimal('4.99'), datetime.datetime(2007, 2, 18, 7, 10, 14, 996577))]

In [127]:
customer.merge(payment, on='customer_id')\
['customer_id first_name last_name email amount payment_date'.split()]\
.sort_values('customer_id')\
.head()

customer_id first_name last_name                          email  amount  \
37            1       Mary     Smith  mary.smith@sakilacustomer.org    4.99   
26            1       Mary     Smith  mary.smith@sakilacustomer.org    0.99   
27            1       Mary     Smith  mary.smith@sakilacustomer.org    4.99   
28            1       Mary     Smith  mary.smith@sakilacustomer.org    0.99   
29            1       Mary     Smith  mary.smith@sakilacustomer.org    0.99   

                 payment_date  
37 2007-04-09 11:52:33.996577  
26 2007-03-02 16:30:04.996577  
27 2007-03-17 11:06:20.996577  
28 2007-03-18 02:25:55.996577  
29 2007-03-19 08:23:42.996577

#### example 3

In [128]:
%%sql
select c.customer_id, first_name, last_name, email, amount, payment_date
from customer c
inner join payment p
on p.customer_id = c.customer_id
where c.customer_id = 2
limit 5;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[(2, 'Patricia', 'Johnson', 'patricia.johnson@sakilacustomer.org', Decimal('2.99'), datetime.datetime(2007, 2, 17, 19, 23, 24, 996577)),
 (2, 'Patricia', 'Johnson', 'patricia.johnson@sakilacustomer.org', Decimal('0.99'), datetime.datetime(2007, 3, 1, 8, 13, 52, 996577)),
 (2, 'Patricia', 'Johnson', 'patricia.johnson@sakilacustomer.org', Decimal('0.99'), datetime.datetime(2007, 3, 2, 0, 39, 22, 996577)),
 (2, 'Patricia', 'Johnson', 'patricia.johnson@sakilacustomer.org', Decimal('5.99'), datetime.datetime(2007, 3, 2, 6, 10, 7, 996577)),
 (2, 'Patricia', 'Johnson', 'patricia.johnson@sakilacustomer.org', Decimal('6.99'), datetime.datetime(2007, 3, 2, 9, 12, 14, 996577))]

In [129]:
customer.merge(payment, on='customer_id')\
['customer_id first_name last_name email amount payment_date'.split()]\
.loc[lambda x: x.customer_id == 2]\
.head()

customer_id first_name last_name                                email  \
47            2   Patricia   Johnson  patricia.johnson@sakilacustomer.org   
48            2   Patricia   Johnson  patricia.johnson@sakilacustomer.org   
49            2   Patricia   Johnson  patricia.johnson@sakilacustomer.org   
50            2   Patricia   Johnson  patricia.johnson@sakilacustomer.org   
51            2   Patricia   Johnson  patricia.johnson@sakilacustomer.org   

    amount               payment_date  
47    2.99 2007-02-17 19:23:24.996577  
48    0.99 2007-03-01 08:13:52.996577  
49    0.99 2007-03-02 00:39:22.996577  
50    5.99 2007-03-02 06:10:07.996577  
51    6.99 2007-03-02 09:12:14.996577

#### example 4

In [130]:
%%sql
select c.customer_id, first_name, last_name, email, amount, payment_date
from customer c
inner join payment p
on p.customer_id = c.customer_id
where first_name like 'A%'
limit 5;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[(29, 'Angela', 'Hernandez', 'angela.hernandez@sakilacustomer.org', Decimal('0.99'), datetime.datetime(2007, 2, 19, 9, 7, 8, 996577)),
 (29, 'Angela', 'Hernandez', 'angela.hernandez@sakilacustomer.org', Decimal('0.99'), datetime.datetime(2007, 2, 19, 10, 10, 46, 996577)),
 (29, 'Angela', 'Hernandez', 'angela.hernandez@sakilacustomer.org', Decimal('7.99'), datetime.datetime(2007, 2, 19, 12, 1, 32, 996577)),
 (29, 'Angela', 'Hernandez', 'angela.hernandez@sakilacustomer.org', Decimal('2.99'), datetime.datetime(2007, 2, 19, 14, 10, 33, 996577)),
 (29, 'Angela', 'Hernandez', 'angela.hernandez@sakilacustomer.org', Decimal('2.99'), datetime.datetime(2007, 2, 19, 17, 46, 35, 996577))]

In [131]:
customer.merge(payment, on='customer_id')\
['customer_id first_name last_name email amount payment_date'.split()]\
.loc[lambda x: x.first_name.str.startswith('A')]\
.head()

customer_id first_name  last_name                                email  \
732           29     Angela  Hernandez  angela.hernandez@sakilacustomer.org   
733           29     Angela  Hernandez  angela.hernandez@sakilacustomer.org   
734           29     Angela  Hernandez  angela.hernandez@sakilacustomer.org   
735           29     Angela  Hernandez  angela.hernandez@sakilacustomer.org   
736           29     Angela  Hernandez  angela.hernandez@sakilacustomer.org   

     amount               payment_date  
732    0.99 2007-02-19 09:07:08.996577  
733    0.99 2007-02-19 10:10:46.996577  
734    7.99 2007-02-19 12:01:32.996577  
735    2.99 2007-02-19 14:10:33.996577  
736    2.99 2007-02-19 17:46:35.996577

![](../images/dvdrental_schema.png)

In [134]:
%%sql
select * from inventory limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(1, 1, 1, datetime.datetime(2006, 2, 15, 10, 9, 17)),
 (2, 1, 1, datetime.datetime(2006, 2, 15, 10, 9, 17))]

In [135]:
%%sql
select * from film limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(133, 'Chamber Italian', 'A Fateful Reflection of a Moose And a Husband who must Overcome a Monkey in Nigeria', 2006, 1, 7, Decimal('4.99'), 117, Decimal('14.99'), 'NC-17', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'chamber':1 'fate':4 'husband':11 'italian':2 'monkey':16 'moos':8 'must':13 'nigeria':18 'overcom':14 'reflect':5"),
 (384, 'Grosse Wonderful', 'A Epic Drama of a Cat And a Explorer who must Redeem a Moose in Australia', 2006, 1, 5, Decimal('4.99'), 49, Decimal('19.99'), 'R', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Behind the Scenes'], "'australia':18 'cat':8 'drama':5 'epic':4 'explor':11 'gross':1 'moos':16 'must':13 'redeem':14 'wonder':2")]

In [141]:
%%sql
select store_id, title 
from inventory i
join film f on i.film_id = f.film_id
limit 6;

 * postgres://postgres:***@localhost:5432/dvdrental
6 rows affected.


[(1, 'Academy Dinosaur'),
 (1, 'Academy Dinosaur'),
 (1, 'Academy Dinosaur'),
 (1, 'Academy Dinosaur'),
 (2, 'Academy Dinosaur'),
 (2, 'Academy Dinosaur')]

In [142]:
inventory.merge(film, on='film_id')[['store_id','title']].head(6)

store_id             title
0         1  Academy Dinosaur
1         1  Academy Dinosaur
2         1  Academy Dinosaur
3         1  Academy Dinosaur
4         2  Academy Dinosaur
5         2  Academy Dinosaur

In [145]:
%%sql
select title, count(title)
from inventory i
inner join film f
on i.film_id = f.film_id
where store_id = 1
group by title
limit 5;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[('Academy Dinosaur', 4),
 ('Affair Prejudice', 4),
 ('Agent Truman', 3),
 ('Airplane Sierra', 2),
 ('Alabama Devil', 3)]

In [153]:
inventory.merge(film, on='film_id')\
.loc[lambda x: x['store_id'] == 1]\
['title']\
.value_counts()\
.sort_index()\
.head()

Academy Dinosaur    4
Affair Prejudice    4
Agent Truman        3
Airplane Sierra     2
Alabama Devil       3
Name: title, dtype: int64

In [154]:
%%sql
select title, count(title) as copies_at_store1
from inventory i
inner join film f
on i.film_id = f.film_id
where store_id = 1
group by title
order by title
limit 5;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[('Academy Dinosaur', 4),
 ('Affair Prejudice', 4),
 ('Agent Truman', 3),
 ('Airplane Sierra', 2),
 ('Alabama Devil', 3)]

In [159]:
(inventory.merge(film, on='film_id')
.loc[lambda x: x['store_id'] == 1]
['title']
.value_counts()
.sort_index()
.reset_index()
.rename(columns={'index': 'title','title':'copies_at_store1'})
.head()
)

title  copies_at_store1
0  Academy Dinosaur                 4
1  Affair Prejudice                 4
2      Agent Truman                 3
3   Airplane Sierra                 2
4     Alabama Devil                 3

In [166]:
## another example

In [167]:
%%sql
select * from film limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(133, 'Chamber Italian', 'A Fateful Reflection of a Moose And a Husband who must Overcome a Monkey in Nigeria', 2006, 1, 7, Decimal('4.99'), 117, Decimal('14.99'), 'NC-17', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'chamber':1 'fate':4 'husband':11 'italian':2 'monkey':16 'moos':8 'must':13 'nigeria':18 'overcom':14 'reflect':5"),
 (384, 'Grosse Wonderful', 'A Epic Drama of a Cat And a Explorer who must Redeem a Moose in Australia', 2006, 1, 5, Decimal('4.99'), 49, Decimal('19.99'), 'R', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Behind the Scenes'], "'australia':18 'cat':8 'drama':5 'epic':4 'explor':11 'gross':1 'moos':16 'must':13 'redeem':14 'wonder':2")]

In [168]:
%%sql
select * from language limit 2;

 * postgres://postgres:***@localhost:5432/dvdrental
2 rows affected.


[(1, 'English             ', datetime.datetime(2006, 2, 15, 10, 2, 19)),
 (2, 'Italian             ', datetime.datetime(2006, 2, 15, 10, 2, 19))]

In [173]:
%%sql
select f.title, l.name as movie_language
from film f
join language l
on l.language_id = f.language_id
limit 5;

 * postgres://postgres:***@localhost:5432/dvdrental
5 rows affected.


[('Chamber Italian', 'English             '),
 ('Grosse Wonderful', 'English             '),
 ('Airport Pollock', 'English             '),
 ('Bright Encounters', 'English             '),
 ('Academy Dinosaur', 'English             ')]

In [172]:
(film.merge(language, on='language_id')
 [['title','name']]
 .rename(columns={'title':'title','name':'movie_language'})
 .head()
)

title        movie_language
0    Chamber Italian  English             
1   Grosse Wonderful  English             
2    Airport Pollock  English             
3  Bright Encounters  English             
4   Academy Dinosaur  English